src/System.Linq/src/System/Linq/Enumerable.cs

```diff
@@ -2266,186 +2620,247 @@ public static decimal Max<TSource>(this IEnumerable<TSource> source, Func<TSourc
 
         public static double Average(this IEnumerable<int> source)
         {
-            if (source == null) throw Error.ArgumentNull("source");
-            long sum = 0;
-            long count = 0;
-            checked
+            Error.CheckSourceNotNull(source);
+            using (IEnumerator<int> e = source.GetEnumerator())
             {
-                foreach (int v in source)
+                if (!e.MoveNext()) throw Error.NoElements();
+                long sum = e.Current;
+                long count = 1;
+                checked
                 {
-                    sum += v;
-                    count++;
+                    while (e.MoveNext())
+                    {
+                        sum += e.Current;
+                        ++count;
+                    }
                 }
+                return (double)sum / count;
```

Comment:
Is this whole change just to avoid the if at the end of the function and one extra addition?  Is there a measurable cost to those?  This transformation doesn't add a ton of complexity, but it does add some, which makes me hesitate if the reward can't be quantified.